TheEngineeringJourney/Assets/Scripts/Map/MapBuilder.cs

```diff
@@ -33,48 +33,76 @@ private void LoadRoomNodeTypeList()
-            
-            // Loop until dungeon successfully built or more than max attempts for node graph
-            while (!dungeonBuildSuccessful && dungeonRebuildAttemptsForNodeGraph <= Settings.MaxMapBuildAttempts)
+            switch (status)
             {
-                // Clear dungeon room game objects and dungeon room dictionary
-                ClearMap(MapBuilderRoomDictionary);
-                
-                dungeonRebuildAttemptsForNodeGraph++;
-
-                // Attempt To Build A Random Dungeon For The Selected room node graph
-                dungeonBuildSuccessful = AttemptToBuildRandomMap(roomNodeGraph, MapBuilderRoomDictionary, roomTemplateList, _roomNodeTypeList);
-            }
-
-
-            if (dungeonBuildSuccessful)
-            {
-                // Instantiate Room Game Objects
-                InstantiateRoomGameObjects(MapBuilderRoomDictionary);
+                case Build.CreateNodeGraph:
+                    (status, rebuildAttemptsForNodeGraph) = AttemptCreating(roomNodeGraph, MapBuilderRoomDictionary, roomTemplateList, _roomNodeTypeList, rebuildAttemptsForNodeGraph);
+                    Debug.Log( $"Attempting creating: {status.ToString()}");
+                    break;
+                case Build.ReAttemptFailedCreateNodeGraph:
+                    // Check status 
+                    // If attempt failed then count up 
+                    status = (int)rebuildAttemptsForNodeGraph == Settings.MaxMapBuildAttempts
+                        ? Build.AttemptFailed
+                        : Build.AttemptBuilding;
+                    break;
+                case Build.AttemptFailed:
+                    // 
+                    status = (int)buildAttempts == Settings.MaxMapBuildAttempts
+                        ? Build.MaxAttemptFailed
+                        : Build.AttemptBuilding;
+                    break;
+                case Build.AttemptBuilding:
+                    // Retry creating node Graph
+                    // Reset rebuildAttemptsForNodeGraph
+                    // Increase attempts
+                    // Set status to try again
+                    roomNodeGraph = currentDungeonLevel.RoomNodeGraphs.SelectRandomRoomNodeGraph();
+                    rebuildAttemptsForNodeGraph = Attempt.Zero;
+                    ++buildAttempts;
+                    status = Build.CreateNodeGraph;
+                    break;
+                case Build.MaxAttemptFailed:
+                    status = Build.Failed; 
+                    break;
+                case Build.Failed:
+                    return Build.Failed;
+                case Build.Success:
+                    InstantiateRoomGameObjects(MapBuilderRoomDictionary);
+                    return Build.Success;
             }
+            Debug.Log(status.ToString());
         }
+        
+    }
+
+    private (Build, Attempt) AttemptCreating(RoomNodeGraphSO roomNodeGraph, Dictionary<string, Room> mapBuilderRoomDictionary, 
```

Comment:
Maybe to little happping here and is too inpure.... 

This in the switch case instead
```
ClearMap(MapBuilderRoomDictionary);
AttemptToBuildRandomMap(roomNodeGraph, MapBuilderRoomDictionary, roomTemplates, _roomNodeTypeList);
++rebuildAttemptsForNodeGraph
```